packages/vue/src/factory.tsx

```diff
@@ -2,18 +2,124 @@
-  type HTMLPolymorphicProps,
-} from '@polymorphic-factory/vue'
+  cloneVNode,
+  defineComponent,
+  h,
+  type AllowedComponentProps,
+  type ComponentCustomProps,
+  type DefineComponent,
+  type ExtractPropTypes,
+  type PropType,
+  type VNodeProps,
+} from 'vue'
+import type { IntrinsicElementAttributes } from './dom.types'
+import { isValidVNodeElement, mergeProps, renderSlotFragments } from './utils'
+
+export type AsChildProps = {
+  asChild?: boolean
+}
+
+export type DOMElements = keyof IntrinsicElementAttributes
+
+export type ElementType = DOMElements | DefineComponent
+
+export type AsChildComponent<
+  Component extends ElementType,
+  P extends Record<string, unknown> = Record<never, never>,
+> = {
+  new (): {
+    $props: AllowedComponentProps &
+      ComponentCustomProps &
+      VNodeProps &
+      ExtractPropTypes<Component> &
+      (Component extends keyof IntrinsicElementAttributes
+        ? IntrinsicElementAttributes[Component]
+        : Record<never, never>) &
+      P & {
+        asChild?: boolean
+      }
+  }
+}
 
-export type HTMLArkComponents = HTMLPolymorphicComponents
+export type HTMLPolymorphicComponents = {
+  [Tag in DOMElements]: AsChildComponent<Tag>
+}
+
+export type HTMLPolymorphicProps<T extends ElementType> = Omit<ExtractPropTypes<T>, 'ref'> & {
+  asChild?: boolean
+}
 
 export type HTMLArkProps<T extends DOMElements> = HTMLPolymorphicProps<T>
 
-/**
- * The ark factory serves as an object of JSX elements to render React components which accept the `as` prop
- */
-export const ark = polymorphicFactory()
+type RenderFunctionArgs = Parameters<typeof h>[0]
+
+export function withAsChild(__component: RenderFunctionArgs) {
+  const component = defineComponent<AsChildProps>(({ asChild }, { attrs, slots }) => {
+    if (!asChild) return () => <__component {...attrs}>{slots.default?.()}</__component>
+    else {
+      return () => {
+        let children = slots.default?.()
+        children = renderSlotFragments(children || [])
+
+        if (Object.keys(attrs).length > 0) {
+          const [firstChild, ...otherChildren] = children
+          if (!isValidVNodeElement(firstChild) || otherChildren.length > 0)
+            // TODO Improve error message
+            throw new Error('Only one child is allowed')
+
+          const mergedProps = mergeProps(firstChild.props ?? {}, attrs)
+          const cloned = cloneVNode(firstChild, mergedProps)
+          // Explicitly override props starting with `on`.
+          // It seems cloneVNode from Vue doesn't like overriding `onXXX` props. So
+          // we have to do it manually.
+          for (const prop in mergedProps) {
+            if (prop.startsWith('on')) {
+              cloned.props ||= {}
+              cloned.props[prop] = mergedProps[prop]
+            }
+          }
+          return cloned
+        }
+
+        if (Array.isArray(children) && children.length === 1) {
+          return children[0]
+        }
+
+        return children
```

Comment:
So if i understand correctly, if there are no attributes on the element calling `asChild` we allow more than one element?